### DISCUSSION TITLE
Using @lit-labs/react in a NextJs app

### DISCUSSION THREAD

**Discussion by martin-dimi** [lit/lit#3376] [Q&A] [answered]
Hey, we've got a library which exposes a suite of Lit components which we'd like to use within our NextJs app. Currently, to get this working, we're simply importing the library client-side and using the custom DOM elements directly:

```
// import the components lib on the whole page so all react components have access to it
function SomePage(...){
  useEffect(() => {
    import('@ourLitPkg/components');
  })

  return (
    <SomeComponent/>
  );
}

// Example of how we use lit-components
export const SomeComponent = (props) => {
  return (
        <our-custom-lit-comopnent color = { props.color } />
  );
```

As handling events/callbacks is not super trivial + adds boilerplate, we wanted to use `@lit-labs/react` to wrap our lit components. From what I saw, it's currently not possible to do this SSR. I tried doing the following:

```
const load = async () => {
  const components = await import("@ourLitPkg/components");
  return createComponent(
    React,
    'our-custom-lit-comopnent',
    components.SomeLITComponent,
    {
      color: 'color',
    }
  );
}
const WrappedComponent = dynamic(load, { ssr: false });

// Example of how we use lit-components
export const SomeComponent = (props) => {
  return (
        <WrappedComponent color = { props.color } />
  );
}
```
This sort of works but it's not passing the color prop correctly. I assume this is pretty far from an actual proper solution too... 

So as the title asks, how would I go about using `@lit-labs/react` in nextjs? Apologies if this is a trivial question...

**Comment by augustjk** [comment, marked as answer]
### No `@lit-labs/react`

You used to have to do the dynamic import in the `useEffect()` for client rendering, due to the necessary DOM APIs.

As of `lit@2.3.0` this is no longer necessary and just directly doing `import '@ourLitPkg/components';` to the top of the file would be fine.

You can still choose the pattern above if you'd like to code split.


### With `@lit-labs/react`

The usage of `next/dynamic` is still necessary since the `@lit-labs/react` package still depends on the presence of `HTMLElement` globally.

The fourth argument to `@lit-labs/react` package's `createComponent` should be an object that maps the React component event handler property name to the name of the event fired by the custom element. So I think the `{ color: 'color' }` is the problem. Any regular props should be handled by the wrapper so you don't need to specifically name them.

Here's a running example: https://stackblitz.com/edit/nextjs-w6vznk?file=pages/index.js
The `name` property here is just a string so isn't specified when using `createComponent`. The custom element dispatches a `'magic'` event though, so we add a `{ onMagic: 'magic' }` to let the wrapper know the prop `onMagic` should be called as a handler for when the `'magic'` event is dispatched.

**Reply by martin-dimi** [reply]
That is very clear - thanks for explaining!

### No `@lit-labs/react`
Simply importing our web components pkg as `import '@ourLitPkg/components';` does not work for us. We're getting:
> ReferenceError: window is not defined  

Hence, why we are lazy loading it. We are using lit `^2.4.0`.

### With `@lit-labs/react`
So basically when using `@lit-labs/react` we need to lazy load the component. Are there any plans of making this more compatible with next?

**Reply by augustjk** [reply]
Hm that's interesting.
Here's a stackblitz with no wrapper, no lazy loading. https://stackblitz.com/edit/nextjs-ssag51?file=pages/index.js

If those components being imported are authored with lit 2.4.0, perhaps there's another dependency that's brought in that might be depending on `window`, or somewhere in the component side-effect. I wonder, if the component code is authored in JS, it might even be `window.customElements.define()`. Replacing that with `globalThis.customElements.define()` or just `customElements.define()` could alleviate that. The stack trace should tell you where `window` is trying to be accessed.

As for `@lit-labs/react`, it is correct that currently it'll only work when lazy loading. We definitely want to support wrapper full compatibility with Next.js, which will also involve actually server-rendering the wrapped components in the server.

**Reply by martin-dimi** [reply]
I see. Our webcomponent library is written in typescript (e.g. we're using decorators for defining the custom element). We're using rollup to bundle it and export it as js (esm) which is what we're importing on nextjs. Maybe our setup there is messed up (simplified version):
```
[
  {
    input: "src/index.ts",
    output: [
      {
        file: "dist/esm/index.js",
        format: "esm",
        sourcemap: true,
      },
    ],
    plugins: [
      resolve(),
      commonjs(),
      typescript({ tsconfig: "./tsconfig.json" }),
      litcss(),
    ],
  },
  {
    input: "dist/esm/index.d.ts",
    output: [{ file: "dist/index.d.ts", format: "esm" }],
    plugins: [dts()],
    external: [/\.(css|less|scss)$/],
  },
]
```

**Reply by augustjk** [reply]
It's likely some other source that introduces `window` then. I'd recommend checking the stack trace to see where it's written and see if it can be removed or escaped by using `isServer` checker mentioned in this section https://lit.dev/docs/ssr/authoring/#browser-only-code. Even if you're just client rendering, Next.js will still load the side-effect-ful import on the server unless you lazy load.
